### ISSUE TITLE
docker entrypoint not waiting for /etc/lightning/lightning-rpc creation

### ISSUE THREAD

**Issue by NicolasDorier** [shesek/spark-wallet#73]
Seems weird, I thought your code was doing it properly, but does not seems to be the case.

```
Using lightningd directory mounted in /etc/lightning...
Starting spark wallet (standalone mode)...
Lightning client connection error { Error: connect ECONNREFUSED /etc/lightning/lightning-rpc
    at Object._errnoException (util.js:992:11)
    at _exceptionWithHostPort (util.js:1014:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect',
  address: '/etc/lightning/lightning-rpc' }
uncaughtException, stopping process
```

**Comment by shesek** [comment]
It does try to wait for it: https://github.com/shesek/spark-wallet/blob/9b500c48c4c32fd61a11995410f17adf9f6cdf8d/scripts/docker-entrypoint.sh#L71-L74

If you're not seeing the "waiting for RPC unix socket..." message printed to STDOUT, this means that the test for `-S $LN_PATH/lightning-rpc` passed successfully (`-S` checks that its a socket file).

Is it possible the `lightning-rpc` file exists, but is not responsive for some other reason?

**Comment by NicolasDorier** [comment]
that's a good question.  I wonder if the file still exist if c-lightning get SIGTERMed.

**Comment by shesek** [comment]
Hmm, yes, this is possible.

The spark server daemon should kill itself on startup if its unable to connect to the rpc socket. Wouldn't your docker-compose setup automatically restart it when that happens, which should get it to work eventually?

**Comment by vswee** [comment]
This isn't isolated to docker. Occurs identically in clean `npm` installation also.
`lightning-cli` commands work for manipulating channels etc. so my assumption is that `lightning-rpc` is operating.

**Comment by NicolasDorier** [comment]
@shesek it still work, but people panic and sometimes NGINX get confused as many containers get created, and does not route to proper backend ip because of this. Then it NGINX need manual restart.

Not big deal, but still a pebble in the shoes.

**Comment by NicolasDorier** [comment]
@shesek another solution which might work is me to delete lightning-rpc when the lightning container restart.

**Comment by shesek** [comment]
I added a connectivity test using `netcat` in 7ce24b3a5e04009b1d0960ccb77cc7652c70e27b. If the unix socket is inaccessible, the docker-entrypoint script will now wait for it to be re-bound before starting the spark server daemon.

This is how starting spark with in inaccessible rpc socket file that is later made available looks like:

```
$ docker run -it -v /tmp/tmp.b04ZnLrVLm/lightning-alice:/etc/lightning spark-standalone
Using lightningd directory mounted in /etc/lightning... nc: unix connect failed: Connection refused
waiting for RPC unix socket... # (hangs here until the file is re-bound)
Starting spark wallet (standalone mode)...
Connected to c-lightning v0.6.2 with id 02619171057a9ff733dc53d44f1713232aa893676b75b416bc76c213324155aa70 on network regtest at /etc/lightning/lightning-rpc

...
```

So still a little error message that shows up, but a less scary one that doesn't crash the thing entirely. :)

Could you build the docker image against master and verify this works?

**Comment by NicolasDorier** [comment]
@shesek can you do as well for ARM?

**Comment by shesek** [comment]
Spark doesn't currently have an ARM image, but it seems like it should be doable with `netcat` similarly to what I did here, or alternatively with `socat`. Is there something you're currently using to run Spark on ARM with Docker?

**Comment by NicolasDorier** [comment]
@shesek well BTCPay currently only support LND on Raspberry. CLightning, while having ARM image is blocked because Spark and Charge are not available on ARM.

**Comment by NicolasDorier** [comment]
anyway, this is a problem for another PR. Let's try with your update for now. Your change with netcat is on latest image?

**Comment by shesek** [comment]
Making an ARM image has been on my todo list, but its indeed for another PR
:)

The change with netcat was still not released in a version, you'll have to
build it manually or wait a little more.

**Comment by NicolasDorier** [comment]
I can wait a bit, thanks!

**Comment by shesek** [comment]
Took a bit longer than I anticipated, but [v0.2.6](https://github.com/shesek/spark-wallet/releases/tag/v0.2.6) is out with this change. Would appreciate if you can confirm this works as expected. Cheers!

**Comment by NicolasDorier** [comment]
@shesek this is separate problem, but because you don't use a solution like `tini`, when I try to restart the docker-compose with a big timeout, the spark container does not get killed.

**Comment by NicolasDorier** [comment]
I can push a PR for this

**Comment by NicolasDorier** [comment]
https://github.com/shesek/spark-wallet/pull/90/files

**Comment by shesek** [comment]
Closing, the original issue reported here is resolved.
